core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala

```diff
@@ -111,3 +116,98 @@ class InternalWriteConverterCtx extends WriteConverterCtx {
     (vector.size until maxVecSize).foreach(idx => put(startIdx + idx, 0.0))
   }
 }
+
+object InternalWriteConverterCtx {
+  def initFrame(keyName: String, names: Array[String]): Unit = {
+    val fr = new water.fvec.Frame(Key.make[Frame](keyName))
+    water.fvec.FrameUtils.preparePartialFrame(fr, names)
+    // Save it directly to DKV
+    fr.update()
+  }
+
+  def finalizeFrame(keyName: String,
+                    res: Array[Long],
+                    colTypes: Array[Byte],
+                    colDomains: Array[Array[String]] = null): H2OFrame = {
+    val fr = DKV.getGet[Frame](keyName)
+    water.fvec.FrameUtils.finalizePartialFrame(fr, res, colDomains, colTypes)
+    validateFrame(fr)
+    new H2OFrame(fr)
+  }
+
+  private def validateFrame(fr: Frame): Unit = {
+
+    if (Log.isLoggingFor("DEBUG")) {
+      if (!fr.vecs().isEmpty) {
+        Log.debug("Number of chunks on frame: " + fr.anyVec.nChunks)
+        val nodes = mutable.Map.empty[H2ONode, Int]
+        (0 until fr.anyVec().nChunks()).foreach { i =>
+          val home = fr.anyVec().chunkKey(i).home_node()
+          if (!nodes.contains(home)) {
+            nodes += (home -> 0)
+          }
+          nodes(home) = nodes(home) + 1
+        }
+        Log.debug("Frame distributed on nodes:")
+        nodes.foreach {
+          case (node, n) =>
+            Log.debug(node + ": " + n + " chunks.")
+        }
+      }
+    }
+
+    // Validate num of chunks in each vector
+    if (!fr.vecs().isEmpty) {
+      val first = fr.vecs.head
```

Comment:
Would it be possible to extract validation logic to a separated method? [MRTask](https://github.com/h2oai/sparkling-water/pull/1451/files#diff-21fbbe6ed7d8cdb3212cdcd8dc0c2bceR193) seems to use something similar.

Reply:
I have tried that, it creates slightly shorter code, but hurts readability.